Support new library logic format

dsaffars opened the pull request:
cherry-picked from https://github.com/b-shi/hipBLASLt/tree/liblogic1

b-shi committed:
Support new library logic format

KKyang commented:
What is the purpose of this PR?

KKyang reviewed (changes requested):
Can you write the main changes in the description? For example, changing list to dict...
Is there any benefits to change data[4] to data["ProblemType"]? Have you tested how much compile time increased/ decreased? And the total size required on disk?

KKyang commented on projects/hipblaslt/tensilelite/Tensile/Utilities/merge.py:
```diff
@@ -49,8 +53,8 @@ def allFiles(startDir):
     return files
 
 def reindexSolutions(data):
-    for i, _ in enumerate(data[5]):
```
This file is deprecated

b-shi replied:
Is `TensileMergeLibrary.py` backward compatible with `merge.py` do they use the same command line arguments?

Just FYI, @babakpst

dsaffars replied:
Reverted in favor of TensileMergeLibrary.py

KKyang commented on projects/hipblaslt/tensilelite/Tensile/Utilities/merge.py:
```diff
@@ -486,22 +564,36 @@ def mergePartialLogics(partialLogicFilePaths, outputDir, forceMerge, trimSize=Tr
     outputFilePath = os.path.join(outputDir, baseFileName)
-    with open(outputFilePath, "w") as outFile:
-        yaml.safe_dump(baseLogicData, outFile, default_flow_style=None)
+    #with open(outputFilePath, "w") as outFile:
```
Remove this

dsaffars replied:
Reverted in favor of TensileMergeLibrary.py

KKyang reviewed (changes requested):
This also breaks ``find_exact.py`` under hipblaslt/Utilities.

b-shi commented:
> Can you write the main changes in the description? For example, changing list to dict... Is there any benefits to change data[4] to data["ProblemType"]? Have you tested how much compile time increased/ decreased? And the total size required on disk?

Changing from list to dict does have some benefits - the main is we can avoid hard-coding the indices. We can add new parameters anywhere without having to add them at the end of the list - this is helpful if we want to group relevant parameters together.

aazz44ss commented:
Can you also update TensileMergeLibrary.py, it's new merge script that we want to replace merge.py

dsaffars committed:
Intro Dictionary based TensileMergeLibrary

	ps1. rebase with rocm repo and fix merge conflicts
	ps2. undo changes to merge.py (being outdated)
	ps3. fix CI issue on LibraryIO.py.

b-shi commented:
@KKyang, @aazz44ss, general question, for the library logics, is it safe to delete solution parameters that are not tunable paramaters and are generated internally in solution structs. For those, do we need them in the logic files?

edit: just realized we need to keep things like soln index, etc..

b-shi commented on projects/hipblaslt/tensilelite/Tensile/LibraryIO.py:
```diff
@@ -456,16 +499,19 @@ def createLibraryLogic(schedulePrefix, architectureName, deviceNames, libraryTyp
+    data = {}
     # Tensile version
-    data.append({"MinimumRequiredVersion": __version__})
+    data["MinimumRequiredVersion"] = __version__
```
@KKyang, do we need to update the version if this PR is merged?

b-shi commented on projects/hipblaslt/tensilelite/Tensile/LibraryIO.py:
```diff
@@ -358,11 +390,10 @@ def parseLibraryLogicList(data, srcFile="?"):
 
     rv = {}
+
+    rv["LibraryLogicVersion"] = "0.0.0"
```
I think we might want to change this `__version__`. I think this field was originally proposed to track lib logic changes.

dsaffars replied:
Removed for now

b-shi reviewed (commented):
@dsaffars, changes look good to me overall. Did you get a chance to test the new merge script changes (maybe also old one if we still need it) to see if the lib logics were converted correctly to dict format?

b-shi commented on projects/hipblaslt/tensilelite/Tensile/LibraryIO.py:
```diff
@@ -456,16 +499,19 @@ def createLibraryLogic(schedulePrefix, architectureName, deviceNames, libraryTyp
     # schedule architecture name and get CU count
     CUCount=getCUCount()
-    data.append({"Architecture": architectureName, "CUCount": CUCount} if architectureName=="gfx942" and CUCount and CUCount!=304 else architectureName)
+    data.append({"ArchitectureName": architectureName, "CUCount": CUCount} if architectureName=="gfx942" and CUCount and CUCount!=304 else architectureName)
```
This needs to be updated to support dict. Its causing issues in CI.
```data = {}
        # Tensile version
        data["MinimumRequiredVersion"] = __version__
        # schedule name
        data["ScheduleName"] = schedulePrefix
        # schedule architecture name and get CU count
        CUCount=getCUCount()
>       data.append({"ArchitectureName": architectureName, "CUCount": CUCount} if architectureName=="gfx942" and CUCount and CUCount!=304 else architectureName)
E       AttributeError: 'dict' object has no attribute 'append'
```

dsaffars replied:
Addressed

aazz44ss commented:
> @KKyang, @aazz44ss, general question, for the library logics, is it safe to delete solution parameters that are not tunable paramaters and are generated internally in solution structs. For those, do we need them in the logic files?
> 
> 
> 
> edit: just realized we need to keep things like soln index, etc..

I think there are a lot of parameters in logic file that are generated by solution.py. Maybe we don't need to keep them in logic file because they will be generated again during build.
